File: src/types/_SkillCheck.ts
import * as DB from "tsondb/schema/dsl"
import { AttributeIdentifier } from "./_Identifier.js"

export const SkillCheck = DB.TypeAlias(import.meta.url, {
  name: "SkillCheck",
  comment: "The attributes’ identifiers of the skill check.",
  type: () =>
    DB.Array(AttributeIdentifier(), {
      minItems: 3,
      maxItems: 3,
    }),
})

export const SkillCheckPenalty = DB.Enum(import.meta.url, {
  name: "SkillCheckPenalty",
  comment: "A specific value that represents a penalty for the associated skill check.",
  values: () => ({
    Spirit: DB.EnumCase({ type: null }),
    HalfOfSpirit: DB.EnumCase({ type: null }),
    Toughness: DB.EnumCase({ type: null }),
    HigherOfSpiritAndToughness: DB.EnumCase({ type: null }),
    SummoningDifficulty: DB.EnumCase({ type: null }),
    CreationDifficulty: DB.EnumCase({ type: null }),
    Object: DB.EnumCase({ type: null }),
  }),
})

File: Sources/OptolithDatabaseSchema/Types/_SkillCheck.swift
import FileDB

/// The attributes’ identifiers of the skill check.
@Embedded
public struct SkillCheck {
  @MinItems(3)
  @MaxItems(3)
  let list: [Attribute.ID]
  }

/// A specific value that represents a penalty for the associated skill check.
@ModelEnum
public enum SkillCheckPenalty {
    case Spirit
    case HalfOfSpirit
    case Toughness
    case HigherOfSpiritAndToughness
    case SummoningDifficulty
    case CreationDifficulty
}
